File: android/app/src/main/java/com/jianxing/browser/sync/HistorySync.kt
package com.jianxing.browser.sync

import android.os.Handler
import android.os.Looper
import com.jianxing.browser.JianXingApp
import com.jianxing.browser.data.HistoryStore
import org.json.JSONArray
import java.util.concurrent.Executors

object HistorySync {
    private val io = Executors.newSingleThreadExecutor()
    private val main = Handler(Looper.getMainLooper())
    private var debounce: Runnable? = null

    fun schedule() {
        if (JianXingApp.instance.accountStore.getSession() == null) return
        debounce?.let { main.removeCallbacks(it) }
        val task = Runnable { syncNow() }
        debounce = task
        main.postDelayed(task, DEBOUNCE_MS)
    }

    fun syncNow(onDone: ((ok: Boolean, changed: Boolean) -> Unit)? = null) {
        val app = JianXingApp.instance
        if (app.accountStore.getSession() == null) {
            onDone?.invoke(true, false)
            return
        }
        io.execute {
            val result = runSync()
            main.post { onDone?.invoke(result.first, result.second) }
        }
    }

    @Synchronized
    private fun runSync(): Pair<Boolean, Boolean> {
        val app = JianXingApp.instance
        val store = app.historyStore
        val client = SyncClient(app.accountStore)
        repeat(2) { attempt ->
            val remote = client.pullHistory()
            if (!remote.ok || remote.entries == null) {
                return false to false
            }
            val remoteEntries = HistoryStore.parseEntries(remote.entries)
            val remoteDeleted = HistoryStore.parseDeletedIds(remote.deletedIds)
            val remoteCleared = remote.clearedAt ?: 0L
            val remoteRevision = remote.revision ?: 0
            val changed = store.mergeFromSync(
                remoteEntries,
                remoteDeleted,
                remoteCleared,
                remoteRevision
            )
            val local = store.exportForSync()
            val remoteSnap = HistoryStore.Snapshot(
                entries = remoteEntries,
                deletedIds = remoteDeleted,
                clearedAt = remoteCleared,
                revision = remoteRevision
            )
            if (HistoryStore.payloadEqual(local, remoteSnap)) {
                store.setRevision(remoteRevision)
                return true to changed
            }
            val entries = JSONArray()
            local.entries.forEach { entries.put(HistoryStore.entryToJson(it)) }
            val deleted = JSONArray()
            local.deletedIds.forEach { deleted.put(it) }
            val pushed = client.pushHistory(
                entries,
                deleted,
                local.clearedAt,
                store.getRevision()
            )
            if (pushed.ok) {
                pushed.revision?.let { store.setRevision(it) }
                return true to true
            }
            val conflict = pushed.error?.contains("先拉取") == true ||
                pushed.error?.contains("已更新") == true
            if (!conflict || attempt == 1) {
                return false to changed
            }
        }
        return false to false
    }

    private const val DEBOUNCE_MS = 8000L
}

File: src/main/history-sync.ts
import { AccountStore } from './account-store';
import {
  HistoryStore,
  historyPayloadEqual,
  type HistoryEntry,
} from './history-store';
import { SyncClient } from './sync-client';

const DEBOUNCE_MS = 8000;

export type HistorySyncResult = {
  ok: boolean;
  error?: string;
  unchanged?: boolean;
  skipped?: boolean;
  revision?: number;
};

export function createHistorySync(options: {
  account: AccountStore;
  store: HistoryStore;
  client: SyncClient;
  onChanged: () => void;
}) {
  const { account, store, client, onChanged } = options;
  let timer: ReturnType<typeof setTimeout> | null = null;
  let inflight: Promise<HistorySyncResult> | null = null;

  async function runSync(): Promise<HistorySyncResult> {
    if (!account.isLoggedIn()) {
      return { ok: true, skipped: true };
    }
    for (let attempt = 0; attempt < 2; attempt += 1) {
      const remote = await client.pullHistory();
      if (!remote.ok) {
        return { ok: false, error: remote.error || '拉取历史记录失败' };
      }
      const merged = store.mergeFromSync(
        (remote.entries || []) as Partial<HistoryEntry>[],
        remote.deletedIds || [],
        remote.clearedAt || 0,
        remote.revision || 0
      );
      const local = store.exportForSync();
      const equal = historyPayloadEqual(local, {
        entries: (remote.entries || []) as HistoryEntry[],
        deletedIds: remote.deletedIds || [],
        clearedAt: remote.clearedAt || 0,
      });
      if (equal) {
        if (typeof remote.revision === 'number') {
          store.setRevision(remote.revision);
        }
        if (merged.changed) onChanged();
        return {
          ok: true,
          unchanged: true,
          revision: remote.revision || store.getRevision(),
        };
      }
      const pushed = await client.pushHistory(
        local.entries,
        local.deletedIds,
        local.clearedAt,
        store.getRevision()
      );
      if (pushed.ok) {
        store.setRevision(pushed.revision || store.getRevision());
        onChanged();
        return {
          ok: true,
          unchanged: false,
          revision: pushed.revision,
        };
      }
      const conflict = /先拉取|已更新/.test(pushed.error || '');
      if (conflict && attempt === 0) continue;
      return { ok: false, error: pushed.error || '上传历史记录失败' };
    }
    return { ok: false, error: '同步历史记录失败' };
  }

  async function syncNow(): Promise<HistorySyncResult> {
    if (inflight) return inflight;
    inflight = runSync().finally(() => {
      inflight = null;
    });
    return inflight;
  }

  function schedule(): void {
    if (!account.isLoggedIn()) return;
    if (timer) clearTimeout(timer);
    timer = setTimeout(() => {
      timer = null;
      void syncNow();
    }, DEBOUNCE_MS);
  }

  return { syncNow, schedule };
}
